src/Mod/PartDesign/App/FeatureRefine.h

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+/****************************************************************************
+ *                                                                          *
+ *   Copyright (c) 2024 <bgbsww@gmail.com>                                  *
+ *                                                                          *
+ *   This file is part of FreeCAD.                                          *
+ *                                                                          *
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+
+#ifndef PARTDESIGN_FeatureRefine_H
+#define PARTDESIGN_FeatureRefine_H
+
+#include "Feature.h"
+
+/// Base class of all features that can be refined PartDesign
+namespace PartDesign
+{
+
+class PartDesignExport FeatureRefine : public PartDesign::Feature
+{
+    PROPERTY_HEADER_WITH_OVERRIDE(PartDesign::FeatureRefine);
+
+public:
+
+    enum class RefineErrorPolicy {
+        Raise = 0,
+        Warn
+    };
+
+    FeatureRefine();
+
+    App::PropertyBool Refine;
+
+protected:
+    //store the shape before refinement
+    TopoShape rawShape;
+
+    bool onlyHasToRefine() const;
+    bool onlyHaveRefined();
```

Comment:
I've got a bit of a problem understanding difference between these two. Can we at least make comment above them to explain difference between them? Maybe then we will be able to come up with better name? Even looking at code I have problems understanding the intent.

Reply:
I've added a comment, but still couldn't come up with a better name. Basically anything sounds sort of ridiculous.

Reply:
I think that my biggest problem with the name is that first method prefixed with `only` is de facto predicate without side effects and other one is action with side effects. I know that there is `const` specifier that should tell us about side-effects (or lack of them), but `const` does not really tell us much as stuff inside properties can still be changed... 

Maybe something like `onlyHasToRefine` and `tryRefineOnly`? `try` suggest that this might fail (i.e. only refine is insufficient) and that it can have side effects. I know that `try` is often used to wrap methods that my throw exception but in that case I cannot think of better solution.

Reply:
@chennes ^ ping? Rest of PR looks good